tasks/uk/census/ons.py

```diff
@@ -0,0 +1,125 @@
+# http://webarchive.nationalarchives.gov.uk/20160105160709/http://ons.gov.uk/ons/guide-method/census/2011/census-data/bulk-data/bulk-data-downloads/index.html
+
+import csv
+import json
+import os
+import re
+import urllib
+import shutil
+from zipfile import ZipFile
+
+from luigi import Task, Parameter, LocalTarget
+
+from tasks.meta import current_session
+from tasks.util import classpath, shell, DownloadUnzipTask, TempTableTask
+
+
+class DownloadUK(Task):
+    API_URL = 'https://www.nomisweb.co.uk/api/v01/dataset/def.sdmx.json?search={}*'
+    DOWNLOAD_URL = 'https://www.nomisweb.co.uk/api/v01/dataset/{id}.bulk.csv?time=2011&measures=20100&geography={geo}'
+
+    table = Parameter()
+
+    def run(self):
+        # Query API, extract table ID from name
+        meta = json.load(urllib.urlopen(self.API_URL.format(self.table)))
+        api_id = (meta['structure']['keyfamilies']['keyfamily'][0]['id']).lower()
+
+        # Download for SA (EW,S) and OA (NI) in a single file
+        with self.output().temporary_path() as tmp:
+            os.mkdir(tmp)
+            with open(os.path.join(tmp, '{}.csv'.format(self.table)), 'w') as outcsv:
+                skip_header = False
+                for geo in ['TYPE258', 'TYPE299']:
+                    remote_file = urllib.urlopen(self.DOWNLOAD_URL.format(id=api_id, geo=geo))
+                    if skip_header:
+                        remote_file.next()
+                    else:
+                        skip_header = True
+                    for l in remote_file:
+                        outcsv.write(l)
+
+    def output(self):
+        return LocalTarget(os.path.join('tmp', classpath(self), self.task_id))
+
+
+class ImportUK(TempTableTask):
+    table = Parameter()
+
+    def requires(self):
+        return DownloadUK(self.table)
+
+    @staticmethod
+    def id_to_column(colid):
+        return re.sub(r'[:/, \-\.\(\)]', '_', '_'.join(colid.split(';')[0].split(':')[-2:]))
```

Comment:
It's better to compile the regexp when you're going to use more than once

Reply:
Never mind https://stackoverflow.com/questions/452104/is-it-worth-using-pythons-re-compile#452143 since 2.5 it compiles and caches it automatically

Reply:
Oh, I didn't know that it compiles automatically. Anyway, it's probably better to extract, for clarity.